src/renderer.js

```diff
@@ -88,18 +97,52 @@ Renderer.prototype.stop = function () {
 };
 
 Renderer.prototype.render = function () {
-  this.resize();
-
   var gl = this.gl;
-  gl.viewport(0, 0, gl.drawingBufferWidth, gl.drawingBufferHeight);
-  gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
+  this.update();
+
+  if (this.updated) {
+    gl.viewport(0, 0, gl.drawingBufferWidth, gl.drawingBufferHeight);
+    gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
 
-  var uniforms = updateUniforms(gl, this.camera, this.glTexture, this.uniforms);
-  renderPass(gl, this.shaders.face, this.buffers.face, uniforms, 'TRIANGLES');
+    renderPass(gl, this.shaders.face, this.buffers.face, this.uniforms, 'TRIANGLES');
+    this.updated = false;
+  }
 
   this.raf = requestAnimationFrame(this.render);
 };
 
+Renderer.prototype.update = function (t) {
+  t = t || Date.now();
+
+  var gl = this.gl;
+  var uniforms = this.uniforms;
+  var camera = this.camera;
+
+  camera.idle(t - 20);
+  camera.flush(t - 100);
+  camera.recalcMatrix(t - 25);
+
+  var updated = this.resize() || !equivalent(camera.computedMatrix, uniforms.u_world);
+
+  if (updated) {
+    var projection = m4.perspective(
+        [],
+        Math.PI/4.0,
+        gl.drawingBufferWidth/gl.drawingBufferHeight,
+        uniforms.u_near,
+        uniforms.u_far
+    );
+
+    uniforms.u_camera = m4.invert([], uniforms.u_view);
+    uniforms.u_world = m4.copy([], camera.computedMatrix);
+    uniforms.u_worldView = m4.multiply([], uniforms.u_view, uniforms.u_world);
+    uniforms.u_worldViewProjection = m4.multiply([], projection, uniforms.u_worldView);
+
+    this.updated = true;
+  }
+  return this;
+};
+
 Renderer.prototype.resize = function (resolution) {
```

Comment:
What I've been doing for other stuff is resizing the image backing the canvas before every render. I.e. canvas.height = canvas.offsetHeight and canvas.width = canvas.offsetWidth. I'm not sure if this is performant, but it seems pretty reasonable. XD